File: c-algorithm/删除排序数组中的重复项_26/删除排序数组中的重复项_26.c
//
// Created by jun on 2020/1/26.
//

/* 参考思想一
其后元素向数组首端移动。注意，CompactIntegers函数需要接受数组及其元素个数作为参数，
函数返回值应为删除操作执行后数组的新元素个数。输出删除后数组中元素的个数并依次输出数组元素。
样例输入: （输入格式说明：5为输入数据的个数，3 4 0 0 2 是以空格隔开的5个整数）
5
3 4 0 0 2
样例输出:（输出格式说明：3为非零数据的个数，3 4 2 是以空格隔开的3个非零整数）
3
3 4 2
样例输入:
7
0 0 7 0 0 9 0
样例输出:
2
7 9
样例输入:
3
0 0 0
样例输出:
0*/

#include <stdio.h>
#include <stdlib.h>

void input(int [], int);

void output(int [], int);

int removeDuplicates(int *nums, int numsSize);

int main(void) {
    int n;
    printf("input number :\n");
    scanf("%d", &n);
    int nums[n];
    input(nums, n);
    int new_length = removeDuplicates(nums, n);
    output(nums, new_length);
    output(nums, n);
    return 0;
}

void output(int nums[], int n) {
    while (n--) {
        printf("%d ", *nums++);
    }
    putchar('\n');
}

int removeDuplicates(int *nums, int numsSize) {
    if (numsSize == 0) {
        return 0;
    }

    int new_index = 0;
    nums[new_index] = nums[0];
    int i;
    for (i = 0; i < numsSize; i++) {
        if (nums[new_index] != nums[i]) {
            nums[++new_index] = nums[i];
        }
    }
    return new_index + 1;
}

void input(int nums[], int n) {
    while (n--) {
        scanf("%d", nums++);
    }
}

File: java-algorithm/src/com/jun/_26删除排序数组中的重复项_26/删除排序数组中的重复项_26.java
package com.jun._26删除排序数组中的重复项_26;

import com.jun.util.ArrayUtil;

/**
 * 给定一个排序数组，你需要在原地删除重复出现的元素，使得每个元素只出现一次，返回移除后数组的新长度。
 * <p>
 * 不要使用额外的数组空间，你必须在原地修改输入数组并在使用 O(1) 额外空间的条件下完成。
 * <p>
 * 示例 1:
 * <p>
 * 给定数组 nums = [1,1,2],
 * <p>
 * 函数应该返回新的长度 2, 并且原数组 nums 的前两个元素被修改为 1, 2。
 * <p>
 * 你不需要考虑数组中超出新长度后面的元素。
 * 示例 2:
 * <p>
 * 给定 nums = [0,0,1,1,1,2,2,3,3,4],
 * <p>
 * 函数应该返回新的长度 5, 并且原数组 nums 的前五个元素被修改为 0, 1, 2, 3, 4。
 * <p>
 * 你不需要考虑数组中超出新长度后面的元素。
 * 说明:
 * <p>
 * 为什么返回数值是整数，但输出的答案是数组呢?
 * <p>
 * 请注意，输入数组是以“引用”方式传递的，这意味着在函数里修改输入数组对于调用者是可见的。
 * <p>
 * 你可以想象内部操作如下:
 * <p>
 * // nums 是以“引用”方式传递的。也就是说，不对实参做任何拷贝
 * int len = removeDuplicates(nums);
 * <p>
 * // 在函数里修改输入数组对于调用者是可见的。
 * // 根据你的函数返回的长度, 它会打印出数组中该长度范围内的所有元素。
 * for (int i = 0; i < len; i++) {
 *     print(nums[i]);
 * }
 *
 * @author Jun
 * @date 2020/1/26 下午10:37
 */
public class 删除排序数组中的重复项_26 {

    public static void main(String[] args) {
        int[] nums = {1, 2, 2, 3, 3, 3, 4, 4, 4, 4};
        int index = removeDuplicates(nums);
        System.out.println(index);
        ArrayUtil.outputArray_1(nums,index);
    }


    public static int removeDuplicates(int[] nums) {
        if (nums.length < 1) {
            return 0;
        }

        int newIndex = 0;
        nums[newIndex] = nums[0];
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] != nums[newIndex]) {
                nums[++newIndex] = nums[i];
            }
        }
        return newIndex + 1;
    }
}